Oxygen_pump and Anestethics_pump 2.0

Crow-Control opened the pull request:
Rewrite of part of the Oxygen_pump code, that let you apply it to other players, as requested by @Ithalan 
It now checks for all head related errors before applying the mask on another player, but i have not yet looked into self-application.

Commented the new code like one should.

Commit:
Added code to use open sprites (+1 squashed commits)

Squashed commits:

[70c3975031] Added open sprites for pumps (+1 squashed commits)

Squashed commits:

[a91ec4aaed] Added pre-apply check, user as argument of can_apply_to_target and send the messages to the target if there is no user. as requested by @PsiOmegaDelta (+1 squashed commits)

Squashed commits:

[218ad5d893] Ready for re-review: Fixed bugs concerning the internals icon straying lid in the HUD, added message when user is connected on destroy and added error when trying to add a tank without the maintanance hatch open. (+1 squashed commits)

Squashed commits:

[cebf9de292] Fixed runtime error when trying to changethe internals hud for mobs without one #17108 (+1 squashed commits)

Squashed commits:

[175b9c4f68] Added two masktypes dedicated to the improved pumps (+1 squashed commits)

Squashed commits:

[e25e6dde9d] Fixed all reviewpoints of @Comma (+1 squashed commits)

Squashed commits:

[8cbf239b85] created a new proc for all checks, moved all checks to that proc, added more checks while in use and changed default presure to constant. (+1 squashed commits)

Squashed commits:

[011a330814] added recent changes to changelog (+1 squashed commits)

Squashed commits:

[2be1aa3127] Finished the basic UI with dedicated template (+1 squashed commits)

Squashed commits:

[6034c2007e] added warning when opening UI, without tank installed (+1 squashed commits)

Squashed commits:

[41c00b5669] Added GUI, right click menu item and Attack-AI action. (+1 squashed commits)

Squashed commits:

[4164058317] added on click removal when attatched and added some redundency with the autoremoval. (+1 squashed commits)

Squashed commits:

[2335f2bb39] Fixed type in changelog (+1 squashed commits)

Squashed commits:

[32298ce523] Added changes to changelog and added maintenance hatch check before mask application (+1 squashed commits)

Squashed commits:

[b0c260f474] Cleanup (+1 squashed commits)

Squashed commits:

[8763ae58a1] Added return on error/if (+1 squashed commits)

Squashed commits:

[7857ca6c41] Fixed indentation error and var and cleanup (+1 squashed commits)

Squashed commits:

[54e6417da7] Mask type variable added, fix for tank warnings and some cleanup (+1 squashed commits)

Squashed commits:

[6971789b49] Moved masktype to variable, fixed maint hatch code, cleaned up empty hand code. (+1 squashed commits)

Squashed commits:

[9da2068992] Maybe this would fix issue #17110 [WIP] (+1 squashed commits)

Squashed commits:

[0e314ec59c] Added the new default on-click action to the changelog (+1 squashed commits)

Squashed commits:

[7dc9d888a4] As advised by @Ithalan, directed on-click to tank UI and removed all mask related code there. (+1 squashed commits)

Squashed commits:

[42e6cd48e0] Whoops i fucked up my commit (+1 squashed commits)

Squashed commits:

[09b7bdf433] Let changelog reflex removal of on-click action and focus on user side changes. As also requested by @Comma (+1 squashed commits)

Squashed commits:

[29359407e8] Notice seems more fitting than warning (+1 squashed commits)

Squashed commits:

[8ae74ec8ae] Removed onclick equip completely, due to being inline with other devices and prevent accidental equip. (+1 squashed commits)

Squashed commits:

[d706a95230] usr vs user? (+1 squashed commits)

Squashed commits:

[9b131bd003] On click action improvements (+1 squashed commits)

Squashed commits:

[c6fd81f253] User vs usr fault on line 61 fixed (+1 squashed commits)

Squashed commits:

[5af2355855] Whoops, note to self: Dont edit with Byond when indentation is at risk. (+1 squashed commits)

Squashed commits:

[4e263c5cee] added changelog (+1 squashed commits)

Squashed commits:

[d29b3d1758] Oxygen_pump checks fix for third party application.

Ithalan commented:
It'd probably be most efficient to put all of the checks in a new proc that you can then just call at all the appropriate times (when starting to put the mask on someone and in attach_mask when it actually tries to put it on a character)

As an alternative to my earlier request for a confirmation when direct clicking, maybe give it the same delay as when putting it on others. That would also make it consistent with when mouse dragging it onto yourself

Crow-Control commented:
@Ithalan calling the proc two times (which does all the checks) or returning to the checks after turning it on, shouldn't be less or more efficient. 

That one is nice, will add that change tonight :)

Ornias1993 commented:
But now i'm reading it, would my code even work with all the returns? :s

Crow-Control commented:
@Ithalan 
I have the commit ready, don't know if it checks out, but I tried to keep the checks as small as possible.
It features the requested delay and checks in the same way as putting it onto others.

I do agree there should be a more general function "check mask placement", but I don't think that should be in these files and I think thats a little byond the scope of this PR and issue.

Crow-Control commented:
@Ithalan I am going to PR it otherwise, after reading into some other comparable devices I conclude the following:
Attaching thing A to player P while they stay seperate objects: should be a drag movement from A to P.
Removing thing B from thing C, while thing C stays a serperate object:  should be a Clicking action
Attatching player P to thing A, while they both stay seperate objects:  should be a Drawing action from P to A.

ergo:
there shouldn't be an on-click attatchment anyway, just like the IV doesnt have an on click "attatch to self" either. I will instead place an "on empty hand" action warning saying: "You need an screwdriver to remove this tank".

comma commented:
That's not what changelog is used for - it's for muggles, not coders. Instead of nitty gritty like that you should just mention player-facing changes there, e.g. these things using drag-on-mob instead of click now.

Crow-Control commented:
@comma  thank you for your feedback.
I was already working on it to be honest, found it to coder-stuffs myself too :)

Crow-Control commented:
With that last addition it should be ready for final review.

Ithalan commented:
#17110 can you verify that the PR fixes this issue?

AI interaction with the wall pump also seems to default to using attack_hand. Might be better to override the ai interaction proc to just display the current tank status of the pump

Ithalan commented on code/game/machinery/oxygen_pump.dm:
```diff
@@ -54,11 +74,8 @@
-		attach_mask(user)
+		// User should drag to equip, warn about removing tank without screwdriver.
+		to_chat(user, "<span class='notice'>You need a screwdriver to remove the tank.</span>")
 	else if(user == breather)
```
It might be better to just remove the entire block above and make clicking on the pump with the maintenance hatch closed just always attempt to open the tank interface, regardless of whether the clicker is also the breather.

The message about the tank down on lines 84-85 (The [contained] has already...etc) is also redundant now that clicking doesn't remove the mask from the pump

Crow-Control replied:
And you are completely right, yet again, that just stupid of me, will do!

Crow-Control commented:
@ithalan issue #17110 is rater vague in reproduction steps, but it should not involve this pr at the moment.  Because i did not touch the removal yet.

And you are completely right, yet again, that just stupid of me, will do!

Crow-Control commented:
Okey, after some testing:
1. retraction on removal #17110  is fixed now.
2. all attack functions don't seem to work

Crow-Control commented:
Okey, @Ithalan 
I looked into it and found some bugs and other mistakes I solved on the way:
1. Oxygenpump now spawns breathmasks instead of gasmasks
2. anastethics pump now spawns medicalmasks instead of gasmasks
3. removed the organic limitation for on-click
4. removed the "put mask back" function, because the masks retracts as soon as you unequip it.
5. Fixed the MAINT stat not working and being unable to replace the tank (old code was simply not working and wrong)
6.  tested  #17110 to the dev and PR code, in both version this didnt NOT happen, this issue can not be repeated. Also if you read the code this simply can not happen. Seems to be an isolated case.
7. Added a Tank check when attaching 
8. Cleaned some code up.

One bug remaining:
- The tank UI can not be opened by anyone other than the mask user, due to the tank needing to be silently equiped before the UI can be opened. This is a big To-Do.

One tweak to think about:
- Preventing some organics to remove the tank (like the AI)

@comma I also would like your input on the current state of the PR.

Crow-Control commented:
Okey, considering the UI:
This seems never to have worked too (like many of the code in this device, it couldn't have worked), it could never be read out by any other than the mask user.

I am now in the proces of making a dedicated Oxygen_pump UI for the oxygen_pump.
It will be placed under the right mouse button menu. Because connecting it to attack-hand could be triggered during surger and you wont need to check the UI much.

For the AI, Ai will get the UI on attackAI, and is on its way.

First UI version would use the Tank template, other template will be done after.

Ithalan commented:
Don't worry about misclicks during surgery opening the UI. there's just as much chance to misclick the patient monitor for the same result and that has never been an issue. Right-click menus are the devil too, so please avoid them.

Crow-Control commented:
@Ithalan agreed, so I did make it disconnect an "attack-hand".

I dont think it is wise to have "show UI" when patient is not attached and "Disconnect" not attached.
The other choice would be to add another (opposit or not) drag, but I personally think it it nicer with a click to disconnect.
I like to keep it in line with other devices: Drag to connect, click to disconnect.

The UI is nothing more than a gimmick you wont use much, that my take on it.

If you see a better way to handle things, please tell me! :)

Crow-Control commented:
**Okey, state of the PR:**
- UI: Check
- AI-UI: Check
- Drag to connect: Check
- Click to disconnect: Check
- Right click to open pump settings: Check
- Checks before attaching: Check
- Timer before attaching: Check
- All known and unknown bugs removed: Check
- Every new and old function tested: Check

It seems ready for final review to me now, there is nothing I can think of that needs to be done now.
All futher improvements can be done in a new Quality of Life PR.

Crow-Control commented:
Squashed all of it as it had became way too many commits by now.

PsiOmegaDelta commented on code/game/machinery/oxygen_pump.dm:
```diff
@@ -1,3 +1,6 @@
+#define TANK_MAX_RELEASE_PRESSURE (3*ONE_ATMOSPHERE)
+#define TANK_DEFAULT_RELEASE_PRESSURE 101
```
Why isn't this ```ONE_ATMOSPHERE```?

Crow-Control replied:
Thats a very good question that should/could also be asked about the tank settings I ripped it from :P 

But al crazy aside, Could/will try out `ONE_ATMOSPHERE` to keep `TANK_DEFAULT_RELEASE_PRESSURE` inline with `TANK_MAX_RELEASE_PRESSURE`.

I will also update the tank settings itself in this PR then, to keep code semi-inline with each other. (and you have a very good point it just looks silly)

Crow-Control replied:
I see where I got it wrong, Tank is default 24 so I just changed the numbers without thinking...

May be the 6  bugs before got to me, serieusly the code was bugged like a wasp-nest :P

PsiOmegaDelta commented on code/game/machinery/oxygen_pump.dm:
```diff
@@ -36,36 +40,62 @@
+		if(!do_mob(usr, target, 25))
+			return
+		// Repeat failure checks.
+		if(!tank || !target || !src || !target.organs_by_name[BP_HEAD] || !target.check_has_mouth() || target.wear_mask || (target.head && (target.head.body_parts_covered & FACE)))
```
Seems it's quite possible to not correctly add/remove/adjust checks before/after the ```do_mob()``` call. 

Safer would be to have a single proc that is called both before and after, and takes a feedback receiver as an argument (to_chat should be able to handle null receivers just fine so wouldn't even have to check that).

Crow-Control replied:
Okey, so just create:
/obj/machinery/oxygen_pump/proc/checks
, Put the checks in there and call it when needed?
That would indeed clean up the code and give chat feedback even when the action is aborted which is now not the case.

Will work on it in a few hours, thanks! :)

PsiOmegaDelta replied:
I'd call it something more telling such as ```/can_apply_to_target()``` or similar, but yes.

Crow-Control replied:
Was more a brainfart just out of bed, but will do :)

Crow-Control commented:
@PsiOmegaDelta I have redone all the checks in the code and added all the checks to one and the same proc that gets called for both the target and the breather.

It seemed that, with the old code, you could remove someones head, without the mask retracting.

Also noticed that you could bug it out by using it on two people at once, so added that check.

comma commented on code/game/machinery/oxygen_pump.dm:
```diff
@@ -36,36 +41,47 @@
-		if(user.wear_mask)
-			to_chat(user, "<span class='notice'>You're already wearing a mask!</span>")
+		// stand still target for 25
+		if(!do_mob(usr, target, 25))
```
traditionally instead of running checks twice, they're put in same condition as do_mob, so you can start regardless. So like
if(!do_mob(usr, target, 25) || !can_apply || can_apply_to_target(target))
Works either way, anyway.

Crow-Control replied:
Thanks @Comma
Didnt know that was possible, will fix!

comma commented on code/game/machinery/oxygen_pump.dm:
```diff
@@ -36,36 +41,47 @@
+
+
+/obj/machinery/oxygen_pump/attack_hand(mob/user as mob)
+	if(stat & MAINT && tank)
```
BYOND can be kinda iffy with order of operations, it's better to enclose bit stuff like that in their own parentheses (same for 'x in list', for future reference).
if((stat & MAINT) && tank)

Crow-Control replied:
Thank you... I think however that is still old crap-code :P

comma commented on code/game/machinery/oxygen_pump.dm:
```diff
@@ -6,10 +9,12 @@
 
 	var/spawn_type = /obj/item/weapon/tank/emergency/oxygen/engi
+	var/mask_type = /obj/item/clothing/mask/breath/
+	var/can_apply = 0
```
This var seems to be unneeded, can use return value of can_apply_to_target proc instead.
In all cases this var is checked, said proc is called anyway.

Crow-Control replied:
Can I just write that as:
(for example)
if(!do_mob(usr, target, 25) || can_apply_to_target(target))?

Crow-Control replied:
Nevermind tested it, that works, thanks!
(if(!do_mob(usr, target, 25) || !can_apply_to_target(target)))

comma commented:
Aside from that - changelog is full of coderside again, just mention changes in click-clickdrag behaviours, the menu and ability to swap tank, and maybe mention that masks/helemets coverage is checked now.
Also shame I couldn't see this earlier, would've probably saved some effort - you could probably just forward attack_hand to the tank inside (or show message if there is none) instead of writing your own menu.

Sbotkin commented:
Can you change the mask's name? So you would 100% know that's the pump's mask.

Crow-Control commented:
@Comma,
Will work at the code tips tomorrow at at once, (22:05 here now), thanks by the way!
I don't agree with the changes from the changelog being coderside.

Considering the changelog:
The device is known to have bugged checks and a non-working UI, and not only coder-side.
People should know they wont have to worry again, even if not a coder.

Considering the UI:
Now that I am done with it, I actually like the idea of an separate UI.
Forwarding it to the UI of the tank was done earlier, but any call to the tank when inside another object (even when use using the proc runner in debug mode ingame), just fails.

@Sbotkin 
I think thats going to be more code than I was hoping to commit. Will look into it, but don't know if I will do...

Crow-Control commented:
Okey, I managed to fix all points of @comma before 22:30.
Enough for now, good luck reviewing the changes ^-^

Crow-Control commented:
@Sbotkin
Must say: I loved your idea...
Mostly because I think it should be actually different masks, like the dropdown emergency oxygen masks are not the same as those used with a tank and the same could be said about an automated wallmounted anesthetic system.

So I couldn't help myself and went into the code again:
Introducing:
The Emergency Mask
and
The Anesthetic Mask

(and honestly someday the Emergency one should be made yellow like in an airplane...whooshh..)

Crow-Control commented:
Squashed and ready for re-review.
@comma @Ithalan @PsiOmegaDelta

PsiOmegaDelta commented on code/game/machinery/oxygen_pump.dm:
```diff
@@ -36,36 +43,40 @@
-	else if(contained.loc == src && !breather)
-		if(user.wear_mask)
-			to_chat(user, "<span class='notice'>You're already wearing a mask!</span>")
+		if(!do_mob(usr, target, 25) || !can_apply_to_target(target))
```
This should be changed into
```
if(!can_apply_to_target(target)) // There is no point in attempting to apply a mask if it's impossible.
	return
usr.visible_message("\The [usr] begins placing the mask onto [target]..")
if(!do_mob(usr, target, 25) || !can_apply_to_target(target))
	return
usr.visible_message("\The [usr] has placed \the mask on [target]'s mouth.")
[etc.]
```

```can_apply_to_target()``` should also be changed to take ```usr``` as an argument instead of using it implicitly (conventionally named ```user```).

Crow-Control replied:
I tried to figure out what you mean with:
**can_apply_to_target() should also be changed to take usr as an argument instead of using it implicitly (conventionally named user).**

You can check current commits, If i done what you suggested the right way.

Ofc the precheck is added now.

Crow-Control replied:
@PsiOmegaDelta 
Can you recheck? changes are done ;)

Ornias1993 commented on code/game/machinery/oxygen_pump.dm:
```diff
@@ -36,36 +43,40 @@
-	else if(contained.loc == src && !breather)
-		if(user.wear_mask)
-			to_chat(user, "<span class='notice'>You're already wearing a mask!</span>")
+		if(!do_mob(usr, target, 25) || !can_apply_to_target(target))
```
I can pass user to can_apply

Can_apply_to_target(target, usr)

And let can_apply send the chat to user instead of usr. But wont chat error out at user is null?

Crow-Control commented:
Changes applied as suggested by @PsiOmegaDelta  and added open sprites while I where at it. :)
also squashed while I was at it.

Ithalan commented:
Looks fine to me now design-wise. Not familiar enough with the code standards to speak for the code changes that Psi requested however

Crow-Control commented:
@Ithalan Thank you, Thats what I wanted you to check... You also play medical frequently so I value your opinion in this mater :)

Crow-Control commented:
Can anyone else handle this instead of Psi?
He said on discord he wasnt feeling well enough to handle his own change request.

@comma or someone else?
